### PR TITLE
Feature: implement choices

### PR DESCRIPTION
**Pull request by hikerpig**
A sketchy implementation of Choices , releated #1166

### CODE REVIEW

**Comment by SirVer** [comment]
Sorry for the slow response. I just came back from vacation and my whole family is down with the flu currently, so I have quite limited time. I will definitely review this as next action on UltiSnips!

File: pythonx/UltiSnips/snippet/parsing/base.py [hunk 1]
```diff
@@ -57,7 +57,10 @@ def _do_parse(parent, text, allowed_tokens):
             else:
                 klass = token_to_textobject.get(token.__class__, None)
                 if klass is not None:
-                    klass(parent, token)
+                    text_object = klass(parent, token)
+
+                    if isinstance(text_object, TabStop):
```
Comment (SirVer) [inline pythonx/UltiSnips/snippet/parsing/base.py]: Add a comment that choices token actually materialize into a TabStop instance, please.

File: pythonx/UltiSnips/snippet/parsing/lexer.py [hunk 2]
```diff
@@ -250,6 +250,41 @@ def __repr__(self):
         return "MirrorToken(%r,%r,%r)" % (self.start, self.end, self.number)
 
 
+class ChoicesToken(Token):
+
+    """${1|o1,o2,o3|}"""
```
Comment (SirVer) [inline pythonx/UltiSnips/snippet/parsing/lexer.py]: Worthwile adding in this comment that choices are taken literally, i.e. no recursive parsing is taking place like it is for other tabstops.

File: pythonx/UltiSnips/text_objects/base.py [hunk 3]
```diff
@@ -143,6 +143,7 @@ def overwrite(self, buf, gtext):
         if self.current_text == gtext:
             return
         old_end = self._end
+
```
Comment (SirVer) [inline pythonx/UltiSnips/text_objects/base.py]: Revert this file?

File: pythonx/UltiSnips/text_objects/choices.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
+#!/usr/bin/env python
+# encoding: utf-8
+
+"""Choices are enumeration values you can choose, by selecting index number,
+it is also a special TabStop
+"""
```
Comment (SirVer) [inline pythonx/UltiSnips/text_objects/choices.py]: empty line after dco comment.

File: pythonx/UltiSnips/text_objects/choices.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+from UltiSnips import vim_helper
+from UltiSnips.position import Position
+from UltiSnips.text_objects.tabstop import TabStop
+
```
Comment (SirVer) [inline pythonx/UltiSnips/text_objects/choices.py]: all UltiSnips import in the same block, please.

File: pythonx/UltiSnips/text_objects/choices.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+        # trying to get what user inputted in current buffer
+        if ctype == "I":
+            self._input_chars.append(cmd_text)
+        elif ctype == "D":
+            line_text = vim_helper.buf[cursor_line - 1]
+            self._input_chars = list(line_text[self._start.col: col])
+
+        inputted_text = "".join(self._input_chars)
+
+        if not len(self._input_chars):
```
Comment (SirVer) [inline pythonx/UltiSnips/text_objects/choices.py]: equivalent: `if not self._input_chars`

File: pythonx/UltiSnips/text_objects/choices.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+                num = int(inputted_text)
+                remained_choice_list = list(self._choice_list)[num - 1: num]
+            else:
+                should_continue_input = True
+        else:
+            remained_choice_list = []
+
+        if should_continue_input:
+            # will wait for further input
+            # maybe we can inform user about this?
```
Comment (SirVer) [inline pythonx/UltiSnips/text_objects/choices.py]: remove? We have no current plans of figuring out the UI for that.

File: test/test_Choices.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    wanted = "1 is a char"
+
+
+class Choices_EmptyChoiceWillBeDiscarded(_VimTest):
+    snippets = ("test", "${1|a,,c|}")
+    keys = "test" + EX
+    wanted = "1.a|2.c"
+
+
+class Choices_WillNotExpand_If_ChoiceListIsEmpty(_VimTest):
```
Comment (SirVer) [inline test/test_Choices.py]: Add a few more tests for the following cases:

- Mirror:`${1|foo,blah|} - $1`
- Non-latin characters: `${1|öääö|trna|some korean?|}`
- Choices as zero tabstop: `${0|foo,blah|}
- Choices as nested element: `${1:${2|foo,blah|}}` - test both overwriting the default text as well as tabbing through to it.

**Comment by SirVer** [comment]
@hikerpig Sorry for the long turnaround, but unfortunately I also caught the tail end of my families sickness and was out for a few days. I did a first pass over the code and I think in general it looks clean, great job! Let me know your thoughts on my comments.

**Comment by hikerpig** [comment]
1. I tend to keep the syntax with precedent implementations, AFAIK, TextMate and vscode, the second one I use most of the time.
2. Your comments are valuable, I will fix the style issues and add more explications in comments.
3. I will try to add some documentation, but I'm very new at developing vim plugin and not very good at writing, this might take some time 🤔  

Will start working on these on the weekend.

File: test/test_Choices.py (hunk 8 above)
Comment (hikerpig) [reply test/test_Choices.py]: I have a question and a problem with these tests.

* Choices as zero tabstop: `${0|foo,blah|}`, what is the expected behavior ? In my test,   `_update()` is called, but `_do_edit()` is not.

  1. So with input `test<tab>`, the result is `1.foo|2.blah`.
  2. So with input `test<tab>1`, the result is `1` .

  If that is correct , should the test named `Choices_AsZeroTabstop_SelectionWillNotWork` ? Emm, looks like a bug.
  
  
* Mirror: `${1|foo,blah|} - $1` , with input `test<tab>1`, the result is `foo - `. Here I see there is indeed a bug, I will need some time to fix this.

**Comment by SirVer** [comment]
1. okay, sounds good to me
2. :)
3. If you feel uncomfortable with this, I can also take a stab at documentation. Just let me know.

File: test/test_Choices.py (hunk 8 above)
Comment (SirVer) [reply test/test_Choices.py]: a) Yes, that is a nasty corner case I have not considered. UltiSnips exits a snippet as soon as `$0` is reached and will not consider any further transformations inside the snippet. In our case this is not expected behavior, because users will expect that the last placeholder still lets them select.... Since the implementation is fairly difficult, I suggest to just state in the documentation that selections are not supported on `$0` and to throw an error if people try it (add a test for the thrown error, please).

b) cool, great that we found a bug through a test i proposed :).

**Comment by hikerpig** [comment]
@SirVer  Sorry for the slow response, now I'm back from the holiday so I basically only have time in the weekend.

I've updated the code, including:

1. Add those tests you mentioned.
2. Tweak the text object implementation to fix the Mirror bug.
3. Throw an error when using choices with `$0`.
4. Fix code style issues, also add some comments.

Unfortunately I'm such a bad writer, so I need your help with the documentation, please 🆘.

**Comment by SirVer** [comment]
Thanks for the updated code! ❤️ 

I am well loaded right now, but I'll review soon and help with the documentation.

**Comment by SirVer** [comment]
I am happy with the code. Let's get this in and I'll add documentation for it once I get around to it.

Thanks for your contribution!
